src/core/net/dnssd_server.cpp

```diff
@@ -113,166 +113,176 @@ void Server::HandleUdpReceive(Message &aMessage, const Ip6::MessageInfo &aMessag
+    for (uint16_t i = 0; i < aRequestHeader.GetQuestionCount(); i++)
     {
-        otLogWarnDns("[server] failed to handle DNS query: %s", otThreadErrorToString(error));
+        NameComponentsOffsetInfo nameComponentsOffsetInfo;
 
-        responseHeader.SetQuestionCount(0);
-        responseHeader.SetAnswerCount(0);
-        responseHeader.SetAdditionalRecordCount(0);
-        responseHeader.SetResponseCode(Header::kResponseServerFailure);
-        IgnoreError(aResponse.SetLength(sizeof(Header)));
-    }
+        VerifyOrExit(OT_ERROR_NONE == Dns::Name::ReadName(aMessage, readOffset, name, sizeof(name)),
+                     response = Header::kResponseFormatError);
+        VerifyOrExit(OT_ERROR_NONE == aMessage.Read(readOffset, question), response = Header::kResponseFormatError);
+        readOffset += sizeof(question);
 
-    aResponse.Write(0, responseHeader);
-}
+        uint16_t qtype = question.GetType();
 
-otError Server::AddQuestionToResponse(const char *    aName,
-                                      const Question &aQuestion,
-                                      Message &       aResponse,
-                                      Header &        aResponseHeader)
-{
-    otError error = OT_ERROR_NONE;
+        VerifyOrExit(qtype == ResourceRecord::kTypePtr || qtype == ResourceRecord::kTypeSrv ||
+                         qtype == ResourceRecord::kTypeTxt || qtype == ResourceRecord::kTypeAaaa,
+                     response = Header::kResponseNotImplemented);
 
-    SuccessOrExit(error = Dns::Name::AppendName(aName, aResponse));
-    SuccessOrExit(error = aResponse.Append(aQuestion));
+        VerifyOrExit(question.GetClass() == ResourceRecord::kClassInternet ||
+                         question.GetClass() == ResourceRecord::kClassAny,
+                     response = Header::kResponseNotImplemented);
 
-    aResponseHeader.SetQuestionCount(1);
+        VerifyOrExit(OT_ERROR_NONE == FindNameComponents(name, compressInfo.GetDomainName(), nameComponentsOffsetInfo),
+                     response = Header::kResponseNameError);
 
-exit:
-    return error;
-}
+        switch (question.GetType())
+        {
+        case ResourceRecord::kTypePtr:
+            VerifyOrExit(nameComponentsOffsetInfo.IsServiceName(), response = Header::kResponseNameError);
+            break;
+        case ResourceRecord::kTypeSrv:
+            VerifyOrExit(nameComponentsOffsetInfo.IsServiceInstanceName(), response = Header::kResponseNameError);
+            resolveAdditional &= ~kResolveAdditionalSrv;
+            break;
+        case ResourceRecord::kTypeTxt:
+            VerifyOrExit(nameComponentsOffsetInfo.IsServiceInstanceName(), response = Header::kResponseNameError);
+            resolveAdditional &= ~kResolveAdditionalTxt;
+            break;
+        case ResourceRecord::kTypeAaaa:
+            VerifyOrExit(nameComponentsOffsetInfo.IsHostName(), response = Header::kResponseNameError);
+            resolveAdditional &= ~kResolveAdditionalAaaa;
```

Comment:
If I understand the code correctly the `resolveAdditional` is used to track which RR types need to be added in additional section vs the ones that are already part of the answer section.

Here we go through the questions, and if see there is question for TXT RR, we clear the flag for TXT in `resolveAdditional`, and then later we won't allow adding of TXT record in additional section.

But how can we know if the questions are all for the same/related resource names?

For example, if there is question for SRV for one name  and AAA for an unrelated host name, I think we still want to include an AAAA record in additional section corresponding to host for the SRV record.

Or if the question contains a PTR question and single SRV question, I think we still need to list all services in response to PTR and add a corresponding SRV record for each in the  additional section except the one that may match the SRV question (and is already included in answer section)?

-----

Maybe before adding any RR in the additional section we add code to check in the previously prepared answer section to see if we already have a matching record (same name and type) there and if so we skip it (i.e. not duplicate it in additional section).  We can use `ResourceRecord::FindRecord()` to find it in answer section. Thoughts?

Reply:
> If I understand the code correctly the `resolveAdditional` is used to track which RR types need to be added in additional section vs the ones that are already part of the answer section.
>
> Here we go through the questions, and if see there is question for TXT RR, we clear the flag for TXT in `resolveAdditional`, and then later we won't allow adding of TXT record in additional section.

You are totally correct. I used this implementation for two reasons:

- It works just fine for almost all DNS-SD client use cases. The DNS-SD client should never add unrelated names in normal use cases. 
- It is very simple to implement, without having to check which RRs have been appended to avoid duplicates. 
- It is in conformance with DNS-SD RFC because addtional RRs are a `SHOULD` requirement but not a `MUST`. It is fine that DNS-SD server do not add additoinal RRs.
- Another thought is that if the DNS-SD client do use unrelated questions, maybe it's exactly its intention **NOT** to have additional RRs. For example, if a client queries a unrated `AAAA` question, it would show that the client have some special intention for the `AAAA` resolution. This (sophisticated) client can certainly also add the related `AAAA` question itself if it wants to. 
 
> But how can we know if the questions are all for the same/related resource names?

We don't know. But they should be in most cases. 

> For example, if there is question for SRV for one name and AAA for an unrelated host name, I think we still want to include an AAAA record in additional section corresponding to host for the SRV record.
>
> Or if the question contains a PTR question and single SRV question, I think we still need to list all services in response to PTR and add a corresponding SRV record for each in the additional section except the one that may match the SRV question (and is already included in answer section)?

Yes, we should better add additional RRs for these two cases. However, this is a `SHOULD` but not `MUST` according to DNS-SD RFC. 

> Maybe before adding any RR in the additional section we add code to check in the previously prepared answer section to see if we already have a matching record (same name and type) there and if so we skip it (i.e. not duplicate it in additional section). We can use `ResourceRecord::FindRecord()` to find it in answer section. Thoughts?

This is a good implementation idea. However, I prefer the current implementation for the  reasons above. 

Thoughts? @abtink

Reply:
I am fine going for the current model now.

I think we can consider enhancing this later.  I feel checking the existing RRs in answer section may not be very complicated to add and it would make the implementation more capable. We can consider it later.

Reply:
> I think we can consider enhancing this later. I feel checking the existing RRs in answer section may not be very complicated to add and it would make the implementation more capable. We can consider it later.

Sure.